Refactor rule and config parsing

radumereuta opened the pull request:
Fixes: #1867
These are incremental changes to the inner parser mostly based on refactoring and doing caching before even getting close to parsing.

Waiting for feedback. Then I can continue with the actual refactoring to scanner generation.

cos commented:
Jenkins: test this please

cos commented on k-distribution/include/builtin/e-kore.k:
```diff
@@ -35,7 +35,7 @@ module OUTER-KORE
 
   syntax KDefinition   ::= KRequireList KModuleList
-                               [klabel(#KDefinition), hook(org.kframework.definition.Definition)]
+                               [klabel(#KDefinition), hook(org.kframework.definition.Definition), latex(\begin{document} #1 #2 \end{document})]
```
@grosu, we need some representation for new lines in latex -- afaik, there was none needed before because the outer was not done at the meta-level.

@radumereuta, could we simply use `\n`?

grosu replied:
Can't you just use conventional latex?

    latex(\begin{document}{#1}\newline{#2} \end{document})

Or are you taking about something else?

radumereuta replied:
@cos, anything you write in that comment, will be put in the latex file. So `\n` in the attribute will be `\n` in latex. I'm not sure what that means in latex.

cos replied:
well, that is the issue, will `\n` be a new line or the `\` and `n` characters? In any case, as long as some form of `\n` is allowed by `e-kore.k`, I can adjust what it means in the latex output.

cos commented:
@radumereuta, the `e-kore.k` definition does not parse correctly the simple definition in `KDocTest`. For some reason it does not reject the "configuration" string when parsing a bubble.

cos commented:
Jenkins: test this please

grosu commented:
@cos: I would say that we can also merge this one as soon as we merge #2198.  Or we can have one of the new students who joined the FSL work on it?

cos commented:
this PR contains changes to the K API and progress towards an API-based KDoc implementation. @andreistefanescu, can you review? 

@grosu, as far as I remember, the latex annotation style is very similar to before. We can make any adjustments when getting back to it after the pending PRs are merged.

andreistefanescu commented:
Looks good at a high level.

radumereuta committed:
upConfiguration

radumereuta committed:
Compact code.

ehildenb committed:
kernel/DefinitionParsing: factor out registerWarnings

radumereuta committed:
Factor in ruleConfigBubbles

in preparation for splitting parsing from syntax generation

radumereuta committed:
Make a separate step for cached sentences

radumereuta committed:
Split parseConfigBubbles into parse and convert

This way we can take advantage of a separate step to do caching and avoid creating extra scanners

radumereuta committed:
Split performParse in two

1. setup scanner
2. parse a bubble

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -269,69 +271,76 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+                  m.att());
+
+            Set<Sentence> configDeclProductions;
+            ParseCache cache = loadCache(gen.getConfigGrammar(module));
```
This and the following line are expensive steps, and you're doing them twice. It would be better if we could avoid redoing this work.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -369,12 +378,61 @@ private Module resolveNonConfigBubbles(Module module, Scanner scanner, RuleGramm
+        return DefinitionTransformer.from(m -> {
+            if (stream(m.localSentences()).noneMatch(s -> s instanceof Bubble && (isRule || ((Bubble) s).sentenceType().equals(configuration))))
+                return m;
+            ParseCache cache = isRule ? loadCache(gen.getRuleGrammar(m)) : loadCache(gen.getConfigGrammar(m));
```
Here's a third place we're calling loadCache

radumereuta replied:
There are now only two places, once when looking for cached bubbles, and another when doing parsing.
I've done some profiling and it seems that the most expensive step is actually `ParseInModule.getExtensionModule()`
`loadCache` didn't even register. I could avoid calling it twice by combining `resolveBubbles` and `resolveCachedBubbles` but it doesn't seem to be a problem, so at the moment I prefer to have them separate for clarity.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -473,36 +545,19 @@ private boolean equalsSyntax(Module _this, Module that) {
     }
 
     private Stream<? extends K> performParse(Map<String, ParsedSentence> cache, ParseInModule parser, Scanner scanner, Bubble b) {
+        parser.setScanner(scanner);
```
I don't understand why there are now two functions here.

radumereuta replied:
Removed and replaced with `resolveBubbles` which computes the minimum amount of scanners.

dwightguth commented on kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java:
```diff
@@ -186,6 +186,10 @@ public Scanner getScanner() {
         return scanner;
     }
 
+    public void setScanner(Scanner scanner) {
```
This seems suspicious to me.

radumereuta replied:
I needed this because I'm now precomputing the scanners before I have all the parsers.
When I get to create the parser, I just lookup for the appropriate scanner and do parsing.

dwightguth reviewed (changes requested):
It definitely looks better than your last stab at refactoring, but I'm concerned about certain relatively expensive steps being executed multiple times now, as well as some oddness surrounding parseBubble vs performParse

radumereuta committed:
Compute minimum amount of scanners

Precompute depending on bubbles found in modules

radumereuta commented:
I did some preliminary testing and here are some numbers:
master: make -j 32 regression-new -> 2m36s
current: make -j 32 regression-new -> 2m32s
With the kserver on.

I ran `kompile -v --backend haskell` over a simple test definition (no kserver), and here are the outputs:
```
$ cat test.k
module TEST
  imports TEST-SYNTAX
  imports INT

  configuration <k> $PGM:Int </k>
  rule 0 => 1
endmodule
```
this branch first time:
```
Outer parsing [57 modules]                                   =  0.751s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.189s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.193s
New scanner: TEST-CONFIG-CELLS                               =  0.123s
Parse configurations [3/3 declarations]                      =  0.636s
New scanner: TEST-RULE-CELLS                                 =  0.120s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.194s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.186s
Parse rules [101/101 rules]                                  =  0.750s
Validate definition                                          =  0.058s
Apply compile pipeline                                       =  0.245s
Kompile to kore                                              =  0.002s
Save to disk                                                 =  0.024s
Backend                                                      =  0.287s
Cleanup                                                      =  0.000s
Total                                                        =  2.765s
```
this branch second time:
```
Outer parsing [57 modules]                                   =  0.725s
Parse configurations [0/3 declarations]                      =  0.270s
Parse rules [0/101 rules]                                    =  0.071s
Validate definition                                          =  0.073s
Apply compile pipeline                                       =  0.290s
Kompile to kore                                              =  0.001s
Save to disk                                                 =  0.022s
Backend                                                      =  0.338s
Cleanup                                                      =  0.000s
Total                                                        =  1.801s
```
master branch first time:
```
Outer parsing [57 modules]                                   =  0.736s
New scanner: TEST-CONFIG-CELLS                               =  0.126s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.173s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.181s
Parse configurations [3/3 declarations]                      =  0.802s
New scanner: TEST-RULE-CELLS                                 =  0.123s
New scanner: STRING-COMMON-RULE-CELLS                        =  0.177s
New scanner: STRING-KORE-RULE-CELLS                          =  0.182s
New scanner: K-IO-RULE-CELLS                                 =  0.248s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.252s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.258s
Parse rules [101/101 rules]                                  =  0.699s
Validate definition                                          =  0.063s
Apply compile pipeline                                       =  0.255s
Kompile to kore                                              =  0.001s
Save to disk                                                 =  0.032s
Backend                                                      =  0.291s
Cleanup                                                      =  0.001s
Total                                                        =  2.890s
```
master branch second time:
```
Outer parsing [57 modules]                                   =  0.714s
New scanner: TEST-CONFIG-CELLS                               =  0.127s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.176s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.185s
Parse configurations [0/3 declarations]                      =  0.738s
New scanner: TEST-RULE-CELLS                                 =  0.119s
New scanner: STRING-COMMON-RULE-CELLS                        =  0.137s
New scanner: STRING-KORE-RULE-CELLS                          =  0.136s
New scanner: K-IO-RULE-CELLS                                 =  0.201s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.200s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.216s
Parse rules [0/101 rules]                                    =  0.473s
Validate definition                                          =  0.081s
Apply compile pipeline                                       =  0.275s
Kompile to kore                                              =  0.001s
Save to disk                                                 =  0.029s
Backend                                                      =  0.315s
Cleanup                                                      =  0.001s
Total                                                        =  2.637s
```

radumereuta committed:
Move verbose message closer to source

dwightguth commented:
I fail to see how the changes you pushed address my concerns with the original pull request. We asked for incremental changes, and adding more non incremental changes onto something that mostly looks good isn't an adequate way to address the concerns that I had with the pull request as it originally existed.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -269,69 +271,76 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+            Module ruleParserModule = gen.getConfigGrammar(m);
+
+            ParseCache cache = loadCache(ruleParserModule);
+            try (ParseInModule parser = RuleGrammarGenerator.getCombinedGrammar(cache.getModule(), isStrict, profileRules, files)) {
```
Here is the first ParseInModule being created. It gets `initialize` and `getScanner` called on it so that the object is fully computed prior to the split into multiple threads, which is important, otherwise the work gets done once on each thread.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -269,69 +271,76 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+
+            Set<Sentence> configDeclProductions;
+            ParseCache cache = loadCache(gen.getConfigGrammar(module));
+            ParseInModule parser = RuleGrammarGenerator.getCombinedGrammar(cache.getModule(), isStrict, profileRules, files);
```
Here a ParseInModule is being created a second time, when before this change, the same ParseInModule was being reused in both places.

radumereuta replied:
This is a code readability choice. The alternatives would be to either collapse two functions together (which make them way too big), or pass parameters to those functions, which would make the parameter list bigger than I would like. I took this decision because generating a `ParseInModule` object is cheap.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -269,69 +271,76 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+                      .parallel()
+                      .filter(s -> s instanceof Configuration)
+                      .map(b -> (Configuration) b)
+                      .flatMap(configDecl -> stream(GenerateSentencesFromConfigDecl.gen(configDecl.body(), configDecl.ensures(), configDecl.att(), parser.getExtensionModule(), kore)))
```
Here getExtensionModule is called in parallel from each thread, which means each thread does the work of computing the extension module, even though we could in theory be using the version computed on line 286.

radumereuta committed:
Compute extensionModule before config desugaring
